lib/Dialect/FIRRTL/Export/FIREmitter.cpp

```diff
@@ -834,8 +836,59 @@ void Emitter::emitStatement(PrintFOp op) {
     ps << "," << PP::space;
     emitExpression(op.getCond());
     ps << "," << PP::space;
-    ps.writeQuotedEscaped(op.getFormatString());
-    for (auto operand : op.getSubstitutions()) {
+
+    // Replace the generic "{{}}" special substitutions with their attributes.
+    // E.g.:
+    //
+    //     "hello {{}} world"(%time)
+    //
+    // Becomes:
+    //
+    //     "hello {{SimulationTime}} world"
+    SmallString<64> formatString;
+    auto origFormatString = op.getFormatString();
+    auto substitutionIdx = 0;
+    SmallVector<Value, 4> substitutions;
+    for (size_t i = 0, e = origFormatString.size(); i != e; ++i) {
+      auto c = origFormatString[i];
+      switch (c) {
+      case '%': {
+        formatString.push_back(c);
+        c = origFormatString[++i];
+        switch (c) {
+        case 'b':
+        case 'd':
+        case 'x':
+          substitutions.push_back(op.getSubstitutions()[substitutionIdx++]);
+          break;
+        default:
+          break;
+        }
+        formatString.push_back(c);
+        break;
+      }
+      case '{':
+        if (origFormatString.slice(i, i + 4) == "{{}}") {
+          formatString.append("{{");
+          FIRRTLTypeSwitch<FIRRTLType>(
+              type_cast<FIRRTLType>(
+                  op.getSubstitutions()[substitutionIdx++].getType()))
+              .Case<FStringTimeType>(
+                  [&](auto) { formatString.append("SimulationTime"); })
+              .Default([&](auto) {
+                emitError(op, "unsupported fstring substitution type");
+              });
+          formatString.append("}}");
+        }
+        i += 3;
+        break;
+      default:
+        formatString.push_back(c);
+      }
+    }
+    ps.writeQuotedEscaped(formatString);
```

Comment:
Thanks for updating the emitter :raised_hands:

Reply:
I use it and like to see it work. 🚀